### PR TITLE
Feature/reorg to anchor height

### PR DESCRIPTION
**Pull request by surangap**
#### What kind of PR is this?:

/kind feature

#### Which issue(s) does this PR fixes?:

Fixes cakedefi/defichain-private#588

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -3010,13 +3010,43 @@ bool CChainState::ActivateBestChainStep(CValidationState& state, const CChainPar
...
+                                    return false;
+                                }
+                            }
+                            else {
+                                //No latest fallbackAnchor, Invalidate blocks back to fallbackCheckpointBlockHeight
+                                if (fallbackCheckpointBlockHeight > 0) { // it doesn't makes sense backward to genesis
+                                    if (!invalidateBlocksTo(checkpointIt->second)) {
+                                        return false;
+                                    }
+                                }
```
Comment (bvbfan) [inline src/validation.cpp]: It looks over-engineering, just make an `uint256 checkpointHash` get it and disconnect blocks at end.

**Review by bvbfan — CHANGES_REQUESTED** [review]

File: src/masternodes/anchors.cpp [hunk 2]
```diff
@@ -780,6 +780,15 @@ bool CAnchorIndex::DbErase(uint256 const & hash)
     return db->Erase(std::make_pair(DB_ANCHORS, hash));
 }
 
+const CAnchorIndex::AnchorRec* CAnchorIndex::GetLatestAnchorUpToDeFiHeight(THeight blockHeightDeFi) const
+{
+    AssertLockHeld(cs_main);
+
+    auto & index = anchors.get<AnchorRec::ByDeFiHeight>();
+    auto it = index.lower_bound(blockHeightDeFi);
+    return (it != index.begin()) ? &(*(--it)) : nullptr;
```
Comment (ShengguangXiao) [inline src/masternodes/anchors.cpp:789]: I think here should compare `it` with `index.end`.

template<typename CompatibleKey>
iterator lower_bound(const CompatibleKey& x)const;
Requires: CompatibleKey is a compatible key of key_compare.
Effects: Returns an iterator pointing to the first element with key not less than x, or end() if such an element does not exist.
Complexity: O(log(n)).

File: src/masternodes/anchors.h [hunk 3]
```diff
@@ -212,12 +212,16 @@ class CAnchorIndex
         // tags for multiindex
         struct ByBtcTxHash{};
         struct ByBtcHeight{};
+        struct ByDeFiHeight{};
+
+        THeight DeFiBlockHeight()const{return anchor.height;}
```
Comment (ShengguangXiao) [inline src/masternodes/anchors.h:217]: Can add some blanks for this line?

**Review by monstrobishi — APPROVED** [review]

File: src/validation.cpp (hunk 1 above)
Comment (surangap) [reply src/validation.cpp]: That's the old implementation. The PR is to use anchor height if it is available, if not back to checkpointHash

Comment (bvbfan) [reply src/validation.cpp]: No, new code is over-engineering. Which is one is more clear
```
CBlockIndex *blockIndex = nullptr;
if (spv::pspv && panchors) { // try anchors first
    auto fallbackAnchor = panchors->GetMostLatestAtDeFiHeight(pindexConnect->nHeight);
    if (fallbackAnchor && (fallbackAnchor->anchor.height >= fallbackCheckpointBlockHeight)) {
        blockIndex = LookupBlockIndex(fallbackAnchor->anchor.blockHash);
    }
}
if (!blockIndex && fallbackCheckpointBlockHeight > 0) { // fallback
    blockIndex = LookupBlockIndex(checkpointIt->second);
}
if (blockIndex) {
    disconnectBlocksTo(blockIndex);
}
```

Comment (surangap) [reply src/validation.cpp]: lesser LOC agreed. but functionality wise it's the same.
will update with a minor change as the following code. thanks
```
CBlockIndex *blockIndex = nullptr;
if (spv::pspv && panchors) { // try anchors first
    auto fallbackAnchor = panchors->GetMostLatestAtDeFiHeight(pindexConnect->nHeight);
    if (fallbackAnchor && (fallbackAnchor->anchor.height > fallbackCheckpointBlockHeight)) {
        blockIndex = LookupBlockIndex(fallbackAnchor->anchor.blockHash);
    }
}
if (!blockIndex && fallbackCheckpointBlockHeight > 0) { // fallback
    blockIndex = LookupBlockIndex(checkpointIt->second);
}
if (blockIndex) {
    disconnectBlocksTo(blockIndex);
}
```

**Review by bvbfan — APPROVED** [review]

**Pull request merged**